pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/partition/BoundedColumnValuePartitionFunction.java

```diff
@@ -0,0 +1,92 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.spi.partition;
+
+import com.google.common.base.Preconditions;
+import java.util.Map;
+
+
+/**
+ * Implementation of {@link PartitionFunction} which partitions based configured column values.
+ *
+ * "columnPartitionMap": {
+ *   "subject": {
+ *     "functionName": "BoundedColumnValue",
+ *     "functionConfig": {
+ *       "columnValues": "Maths|English|Chemistry"
+ *     }
+ *   }
+ * }
+ * With this partition config on column "subject", partitionId would be 1 for Maths, 2 for English and so on.
+ * partitionId would be "0" for all other values which is not configured but may occur.
+ */
+public class BoundedColumnValuePartitionFunction implements PartitionFunction {
+  private static final int DEFAULT_PARTITION_ID = 0;
+  private static final String NAME = "BoundedColumnValue";
+  private static final String COLUMN_VALUES = "columnValues";
+  private final Map<String, String> _functionConfig;
+  private final String[] _values;
```

Comment:
Make it a map to accelerate the lookup?

Reply:
I kept it array because arrays size would not be much and HashMap would only add extra memory overhead. I can update it to use Map if you think that would be faster even with small size of array :)

Reply:
Under ~10 values linear search in an array tends to be faster anyway (much more than 10 values when equality is cheaper to verify than strings)

Reply:
Sounds good, let's keep array then